content/blog/2021/06/writing-your-first-policy-with-kubewarden.md

```diff
@@ -0,0 +1,554 @@
+goal of this post is not to record or audit requests :)
+
+```json
+{
+	"kind": "AdmissionReview",
+	"apiVersion": "admission.k8s.io/v1",
+	"request": {
+		"uid": "150d1761-43fb-4f36-bb73-7a3888a0bca2",
+		"kind": {
+			"group": "",
+			"version": "v1",
+			"kind": "Namespace"
+		},
+		"resource": {
+			"group": "",
+			"version": "v1",
+			"resource": "namespaces"
+		},
+		"requestKind": {
+			"group": "",
+			"version": "v1",
+			"kind": "Namespace"
+		},
+		"requestResource": {
+			"group": "",
+			"version": "v1",
+			"resource": "namespaces"
+		},
+		"name": "test-policy-ns",
+		"namespace": "test-policy-ns",
+		"operation": "CREATE",
+		"userInfo": {
+			"username": "kubernetes-admin",
+			"groups": ["system:masters", "system:authenticated"]
+		},
+		"object": {
+			"kind": "Namespace",
+			"apiVersion": "v1",
+			"metadata": {
+				"name": "test-policy-ns",
+				"uid": "2cc6b1fc-7896-4091-b9c3-e034c89c0535",
+				"creationTimestamp": "2021-05-27T15:55:39Z",
+				"managedFields": [{
+					"manager": "kubectl",
+					"operation": "Update",
+					"apiVersion": "v1",
+					"time": "2021-05-27T15:55:39Z",
+					"fieldsType": "FieldsV1",
+					"fieldsV1": {
+						"f:status": {
+							"f:phase": {}
+						}
+					}
+				}]
+			},
+			"spec": {
+				"finalizers": ["kubernetes"]
+			},
+			"status": {
+				"phase": "Active"
+			}
+		},
+		"oldObject": null,
+		"dryRun": false,
+		"options": {
+			"kind": "CreateOptions",
+			"apiVersion": "meta.k8s.io/v1"
+		}
+	}
+}
+```
+
+I save this request as `request.json`. Now, we have all the
+ingredients to dry run the policy, so let's do it:
+
+```console
+$ kwctl run policy.wasm --request-path request.json 2> /dev/null | jq
+{
+  "uid": "150d1761-43fb-4f36-bb73-7a3888a0bca2",
+  "allowed": true
+}
+```
+
+As expected, the response of the policy is that the request is
+accepted. This is a really good start! Now, let's go and write the
+real policy logic.
+
+## Policy implementation
+
+We have to implement three main checks based on the restrictions that
+have been described by our compliance team:
+
+1. The namespace name has to meet a specific prefix format
+   (`t-\w+-\w+` -- e.g. `t-ei-billing`, `t-sre-security`...)
+1. Two annotations have to exist:
+    1. `compliance.my-company.com/team-contact-email`: has to be an
+    email address with the host part pointing to `@my-company.com`; we
+    won't allow other domain names as contact email.
+    1. `compliance.my-company.com/team-region`: has to be one (and
+only one) of `AMER`, `APAC` or `EMEA`.
+
+Let's implement the first check:
+
+```go
+package main
+
+import (
+	"fmt"
+	"github.com/kubewarden/gjson"
+	kubewarden "github.com/kubewarden/policy-sdk-go"
+	"regexp"
+	wapc "github.com/wapc/wapc-guest-tinygo"
+)
+
+func main() {
+	wapc.RegisterFunctions(wapc.Functions{
+		"validate": func(payload []byte) ([]byte, error) {
+			namespaceName := gjson.GetBytes(payload, "request.object.metadata.name")
```

Comment:
Please comment the code !

Reply:
I was thinking this code was self-explanatory. What kind of comments do you expect here? The question is honest, I think comments that basically describe what the logic does don't have a lot of value, that's why I tend to favour self-explanatory code if possible. However, given your comment it seems this code is not self-explanatory, so there must be something to improve here...

Reply:
I think the whole code is pretty self-explanatory. **Maybe** the only line to "explain" a bit is the `gjson.GetBytes`.

This could be covered with something like:
```
// extract the name of the namespace by looking into the resource metadata attribute
```